### PR TITLE
Add missing TS definition for ServerOptions.genReqId function arg

### PR DESCRIPTION
**Pull request by Drakhart**
## tl;dr
This pull request adds a missing Typescript definition for the `ServerOptions.genReqId` function's request argument, which renders the function unusable for processing the request under Typescript.

## Motivation
I was trying to use the incoming request inside `genReqId` to parse the `x-apigateway-context` header and use the request id provided by the AWS API Gateway service, falling back to a newly generated UUID v4 if not present, but then realized that Typescript was complaining whenever I added the request argument.

## Investigation details
According to both [the documentation](https://github.com/fastify/fastify/blob/master/docs/Server.md#genreqid) and [the codebase](https://github.com/fastify/fastify/blob/master/lib/reqIdGenFactory.js#L6), this function receives the incoming request as an argument, but its type definition is missing in [the namespace](https://github.com/fastify/fastify/blob/master/fastify.d.ts#L214) and hence trying to use it in Typescript produces an overload error.

**Example code used to reproduce the error:**
```
import * as fastify from 'fastify'
import { IncomingMessage, ServerResponse } from 'http'

const app = fastify({
  genReqId: (req: fastify.FastifyRequest): string => {
    return 'foo'
  }
})

app.get('/', async (req: fastify.FastifyRequest<IncomingMessage>, res: fastify.FastifyReply<ServerResponse>) => {
  res.send('Hello, World!')
})

app.listen(3000)
```

**Error message:**
```
src/index.ts:5:3 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Type '(req: fastify.FastifyRequest<IncomingMessage, fastify.DefaultQuery, fastify.DefaultParams, fastify.DefaultHeaders, any>) => string' is not assignable to type '() => string | number'.

5   genReqId: (req: fastify.FastifyRequest): string => {
    ~~~~~~~~

  ../../git/fastify/fastify.d.ts:214:5
    214     genReqId?: () => number | string,
            ~~~~~~~~
    The expected type comes from property 'genReqId' which is declared here on type 'ServerOptionsAsSecureHttp2'
  ../../git/fastify/fastify.d.ts:20:18
    20 declare function fastify(opts?: fastify.ServerOptionsAsSecureHttp2): fastify.FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>;
                        ~~~~~~~
    The last overload is declared here.


Found 1 error.
```

## Outcome
After adding the missing type definition Typescript compiles just fine and `genReqId` accepts the request argument.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Drakhart** [comment]
I'm not sure why the TAP unit tests in `test/close.test.js` are failing in Mac.  They seem unrelated to these PR changes, plus I've run the same test suit in my MacBook and everything went fine.

Just to save you guys the hassle of looking for it, this is GitHub CI's output related to the error ([taken from here](https://pipelines.actions.githubusercontent.com/IW4E0xAhna4DVZwimUT6L7j3gyacayN9wFr7p7uURBqpuOPUNe/_apis/pipelines/1/runs/733/signedlogcontent/8?urlExpires=2020-02-02T18%3A24%3A21.5696875Z&urlSigningMethod=HMACV1&urlSignature=hsT8t3WlVY5XOMlf4RPnqntHSczDgoleW9OI2%2BRMWGk%3D)):
```
2020-02-02T18:18:14.9896990Z not ok 8 - test/close.test.js # time=121836.729ms
2020-02-02T18:18:14.9897640Z   ---
2020-02-02T18:18:14.9897780Z   timeout: 240000
2020-02-02T18:18:14.9897920Z   file: test/close.test.js
2020-02-02T18:18:14.9898020Z   childId: 7
2020-02-02T18:18:14.9898230Z   command: /Users/runner/hostedtoolcache/node/12.14.1/x64/bin/node
2020-02-02T18:18:14.9898350Z   args:
2020-02-02T18:18:14.9898810Z     - test/close.test.js
2020-02-02T18:18:14.9898940Z   stdio:
2020-02-02T18:18:14.9899340Z     - 0
2020-02-02T18:18:14.9899780Z     - pipe
2020-02-02T18:18:14.9900310Z     - 2
2020-02-02T18:18:14.9900490Z   cwd: /Users/runner/runners/2.164.0/work/fastify/fastify
2020-02-02T18:18:14.9900610Z   exitCode: 1
2020-02-02T18:18:14.9900890Z   ...
2020-02-02T18:18:14.9905620Z {
2020-02-02T18:18:14.9907130Z     # Subtest: close callback
2020-02-02T18:18:14.9907260Z         1..4
2020-02-02T18:18:14.9908040Z         ok 1 - should not error
2020-02-02T18:18:14.9909140Z         ok 2 - type is [object Object]
2020-02-02T18:18:14.9909610Z         ok 3 - should not error
2020-02-02T18:18:14.9910090Z         ok 4 - expect truthy value
2020-02-02T18:18:14.9910570Z     ok 1 - close callback # time=65.916ms
2020-02-02T18:18:14.9910820Z     
2020-02-02T18:18:14.9912010Z     # Subtest: inside register
2020-02-02T18:18:14.9912190Z         1..5
2020-02-02T18:18:14.9913000Z         ok 1 - should not error
2020-02-02T18:18:14.9913940Z         ok 2 - expect truthy value
2020-02-02T18:18:14.9914580Z         ok 3 - should be equal
2020-02-02T18:18:14.9915070Z         ok 4 - should not error
2020-02-02T18:18:14.9915570Z         ok 5 - expect truthy value
2020-02-02T18:18:14.9916070Z     ok 2 - inside register # time=10.444ms
2020-02-02T18:18:14.9916240Z     
2020-02-02T18:18:14.9916720Z     # Subtest: close order
2020-02-02T18:18:14.9917020Z         1..5
2020-02-02T18:18:14.9918260Z         ok 1 - should not error
2020-02-02T18:18:14.9918830Z         ok 2 - should be equal
2020-02-02T18:18:14.9919320Z         ok 3 - should be equal
2020-02-02T18:18:14.9919800Z         ok 4 - should not error
2020-02-02T18:18:14.9920260Z         ok 5 - should be equal
2020-02-02T18:18:14.9921160Z     ok 3 - close order # time=20.506ms
2020-02-02T18:18:14.9921320Z     
2020-02-02T18:18:14.9921570Z     # Subtest: should not throw an error if the server is not listening
2020-02-02T18:18:14.9921990Z         1..2
2020-02-02T18:18:14.9922760Z         ok 1 - type is [object Object]
2020-02-02T18:18:14.9923540Z         ok 2 - should not error
2020-02-02T18:18:14.9924310Z     ok 4 - should not throw an error if the server is not listening # time=3.298ms
2020-02-02T18:18:14.9924470Z     
2020-02-02T18:18:14.9924970Z     # Subtest: onClose should keep the context
2020-02-02T18:18:14.9925310Z         1..4
2020-02-02T18:18:14.9926140Z         ok 1 - expect truthy value
2020-02-02T18:18:14.9926840Z         ok 2 - expect truthy value
2020-02-02T18:18:14.9927570Z         ok 3 - should be equal
2020-02-02T18:18:14.9928260Z         ok 4 - should not error
2020-02-02T18:18:14.9928810Z     ok 5 - onClose should keep the context # time=4.204ms
2020-02-02T18:18:14.9928960Z     
2020-02-02T18:18:14.9929980Z     # Subtest: Should return error while closing - injection
2020-02-02T18:18:14.9930350Z         1..4
2020-02-02T18:18:14.9930930Z         ok 1 - should not error
2020-02-02T18:18:14.9931540Z         ok 2 - should be equal
2020-02-02T18:18:14.9932030Z         ok 3 - expect truthy value
2020-02-02T18:18:14.9932630Z         ok 4 - should be equal
2020-02-02T18:18:14.9933430Z     ok 6 - Should return error while closing - injection # time=287.437ms
2020-02-02T18:18:14.9933590Z     
2020-02-02T18:18:14.9934900Z     # Subtest: Current opened connection should continue to work after closing and return "connection: close" header - return503OnClosing: false
2020-02-02T18:18:14.9935420Z         ok 1 - should not error
2020-02-02T18:18:14.9935910Z         ok 2 - should match pattern provided
2020-02-02T18:18:14.9936410Z         ok 3 - should match pattern provided
2020-02-02T18:18:14.9937030Z         ok 4 - should match pattern provided
2020-02-02T18:18:14.9937670Z         ok 5 - should match pattern provided
2020-02-02T18:18:14.9937830Z         1..5
2020-02-02T18:18:14.9938600Z     ok 7 - Current opened connection should continue to work after closing and return "connection: close" header - return503OnClosing: false # time=24.064ms
2020-02-02T18:18:14.9938750Z     
2020-02-02T18:18:14.9939380Z     # Subtest: Current opened connection should not accept new incoming connections
2020-02-02T18:18:14.9940090Z         ok 1 - should not error
2020-02-02T18:18:14.9940850Z         not ok 2 - test unfinished
2020-02-02T18:18:14.9941460Z           ---
2020-02-02T18:18:14.9941600Z           stack: |
2020-02-02T18:18:14.9941980Z             Object.<anonymous> (test/close.test.js:187:3)
2020-02-02T18:18:14.9942850Z             Module.replacementCompile (node_modules/append-transform/index.js:58:13)
2020-02-02T18:18:14.9943460Z             Object.<anonymous> (node_modules/append-transform/index.js:62:4)
2020-02-02T18:18:14.9943970Z             Object.<anonymous> (node_modules/nyc/bin/wrap.js:27:4)
2020-02-02T18:18:14.9944140Z             internal/main/run_main_module.js:17:11
2020-02-02T18:18:14.9944340Z           test: Current opened connection should not accept new incoming connections
2020-02-02T18:18:14.9944480Z           at:
2020-02-02T18:18:14.9944840Z             line: 187
2020-02-02T18:18:14.9945300Z             column: 3
2020-02-02T18:18:14.9945690Z             file: test/close.test.js
2020-02-02T18:18:14.9946090Z             function: Object.<anonymous>
2020-02-02T18:18:14.9946450Z           source: >
2020-02-02T18:18:14.9947340Z             t.test('Current opened connection should not accept new incoming connections',
2020-02-02T18:18:14.9947780Z             t => {
2020-02-02T18:18:14.9948040Z           ...
2020-02-02T18:18:14.9952030Z         
2020-02-02T18:18:14.9952230Z         1..2
2020-02-02T18:18:14.9952380Z         # failed 1 of 2 tests
2020-02-02T18:18:14.9953890Z     not ok 8 - Current opened connection should not accept new incoming connections # time=120040.811ms
2020-02-02T18:18:14.9954060Z     
2020-02-02T18:18:14.9954520Z     1..8
2020-02-02T18:18:14.9955130Z     # failed 1 of 8 tests
2020-02-02T18:18:14.9955490Z     # time=120484.6ms
2020-02-02T18:18:14.9956320Z }
```

**Comment by jsumners** [comment]
Please don't request review from me for TS stuff. I do not know TS.

**Comment by Drakhart** [comment]
Sorry about that, James, GitHub suggested that you had reviewed those files recently.  I'll keep that in mind for future contributions.

**Comment by Ethan-Arrowood** [comment]
Sometimes the CI breaks and we just have to rerun it. I'll let someone else from the team review that as they know more about it than I do.

**Comment by Drakhart** [comment]
Thank you, Ethan.
